### PR TITLE
nrf_cloud_fota_poll: Add possibility to be non-blocking

### PR DESCRIPTION
**Pull request by jtguggedal**
Add the option to use nrf_cloud_fota_poll_process() in a
non-blocking manner.
The current, blocking behavior is maintained, and the async
option is enabled by providing a callback function for error
handling.

### CODE REVIEW

File: include/net/nrf_cloud_fota_poll.h [hunk 1]
```diff
@@ -30,18 +30,55 @@ enum nrf_cloud_fota_reboot_status {
...
-	bool full_modem_fota_supported;
+
+	/** Optional, user-provided callback function to handle errors.
+	 *  If the function is provided, @ref nrf_cloud_fota_poll_process will be non-blocking and
+	 *  the user will receive error events asynchronously.
+	 *
+	 * If the function is not provided, @ref nrf_cloud_fota_poll_process will be blocking and
+	 * return an error code when an error occurs.
+	 */
+	fota_error_handler_t error_fn;
```
Comment (jayteemo) [inline include/net/nrf_cloud_fota_poll.h:81]: if an error does occur is the user responsible for reporting the error status to the cloud?
they would need the job id.

Comment (jtguggedal) [reply include/net/nrf_cloud_fota_poll.h:81]: I think this should be handled now, I added a subsequent call to `update_job_status()` the places where `error_fn()` is called

File: subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_fota [hunk 2]
```diff
@@ -45,6 +40,12 @@ source "subsys/logging/Kconfig.template.log_config"
 
 endif # NRF_CLOUD_FOTA
 
+config NRF_CLOUD_FOTA_DOWNLOAD_FRAGMENT_SIZE
+	int "Fragment size for FOTA downloads"
+	depends on FOTA_DOWNLOAD
+	range 128 1900
+	default 1900
```
[the remark this answers is not in the document]
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/Kconfig.nrf_cloud_fota]: 1900 doesn't leave a lot of margin, so I lowered the default back to 1700, but keep the range up to 1900.
That's also why I added the option to set it at runtime, to be able to add some sort of logic to optimize this on device

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c [hunk 3]
```diff
@@ -198,6 +218,26 @@ static void process_pending_job(struct nrf_cloud_fota_poll_ctx *ctx)
...
+							   struct nrf_cloud_fota_poll_ctx,
+							   timeout_work);
+
+	LOG_ERR("Timeout; FOTA download took longer than %d minutes", CONFIG_FOTA_DL_TIMEOUT_MIN);
+
+	fota_status = NRF_CLOUD_FOTA_TIMED_OUT;
+	fota_status_details = FOTA_STATUS_DETAILS_TIMEOUT;
+
+	/* Job was not successful, send status to nRF Cloud */
+	/* TODO: Should not be done on system work queue */
```
[the remark this answers is not in the document]
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c]: I updated this now, and think it should be okay now.
Now it just does a call to `fota_download_cancel()`, which is non-blocking. Then the event is sent from the `fota_download` callback instead, which should be fine.

**Review by jayteemo — APPROVED** [review]

**Review by plskeggs — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c]: https://github.com/nrfconnect/sdk-nrf/pull/17022

[the remark this answers is not in the document]
Comment (jtguggedal) [reply subsys/net/lib/nrf_cloud/src/nrf_cloud_fota_poll.c]: Thanks for following up on this!

File: doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst [hunk 4]
```diff
@@ -659,6 +659,13 @@ Libraries for networking
 
   * Added support for dictionary logs using REST.
 
+* :ref:`lib_nrf_cloud_fota` library:
+
+  * Made the :kconfig:option:`CONFIG_NRF_CLOUD_FOTA_DOWNLOAD_FRAGMENT_SIZE` Kconfig option available and used also when :kconfig:option:`CONFIG_NRF_CLOUD_FOTA_POLL` is enabled.
+    The range of the option is now 128-1900 bytes, and the default value is 1700 bytes.
+
+  * The function :c:func:`nrf_cloud_fota_poll_process` can now be used asynchrounously if a callback to handle errors is provided.
```
Comment (divipillai) [inline doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:667]:
```suggestion
* Updated:

  * The :kconfig:option:`CONFIG_NRF_CLOUD_FOTA_DOWNLOAD_FRAGMENT_SIZE` Kconfig option is made available and used also when the :kconfig:option:`CONFIG_NRF_CLOUD_FOTA_POLL` Kconfig option is enabled.
    The range of the option is now from 128 to 1900 bytes, and the default value is 1700 bytes.
  * The function :c:func:`nrf_cloud_fota_poll_process` can now be used asynchrounously if a callback to handle errors is provided.
```

File: include/net/nrf_cloud_fota_poll.h [hunk 5]
```diff
@@ -30,18 +30,55 @@ enum nrf_cloud_fota_reboot_status {
...
+	struct k_work_delayable timeout_work;
+	bool is_nonblocking;
+	bool full_modem_fota_supported;
 	const char *device_id;
+
+	/* Public variables */
+
+	/** HTTP fragment size. Sets the size of the requested payload in each HTTP request when
+	 *  downloading the firmware image. Note that the header size is not included in this size.
+	 *  It is imporant that the total size of each fragment (HTTP headers + payload) fits in the
```
Comment (divipillai) [inline include/net/nrf_cloud_fota_poll.h:62]:
```suggestion
	 *  It is important that the total size of each fragment (HTTP headers + payload) fits in the
```

**Review by divipillai — APPROVED** [review]

**Review by glarsennordic — APPROVED** [review]

**Review by tony-le-24 — APPROVED** [review]

**Pull request merged**